Proxy X11 Forwarding Support

fspmarshall opened the pull request:
Support for X11 forwarding to `openssh` nodes ~(WIP)~:

- [x] Basic forwarding logic to allow recording proxy to correctly handle X11 forwarding requests and channels.
- [x] Make X11 forwarding controllable via RBAC.
- [x] Support the `permit-x11-forwarding` cert extension.
- [x] Emit audit events for x11 forwarding requests & channels (recording proxy only).

Closes #3401

---

This PR enables minimal X11 forwarding support for teleport clusters by adding the ability to control the `permit-X11-forwarding` certificate extension via RBAC and giving teleport proxies the ability to correctly handle X11 forwarding requests/channels when in recording mode (X11 forwarding technically already worked when not in recording mode, but wasn't particularly useful without `permit-X11-forwarding` extension support).  In order to leverage X11 forwarding, the user must hold a role with the `permit_x11_forwarding` option enabled:

```yaml
kind: role
metadata:
  # ...
spec:
  options:
    permit_x11_forwarding: true
  # ...
version: v3
```

Since this PR only adds support to the proxy, `openssh` still needs to be used for both the client and node.  Assuming `node.example.com` were an `openssh` node and `alice` had `permit_x11_forwarding: true`, she could graphically edit the contents of `example.txt` like so:

```bash
$ ssh -X -J proxy.example.com alice@node.example.com gvim example.txt
```

Which, if the proxy is running in recording mode, would produce an audit event something like this:

```
{
    "event": "x11",
    "code": "T3008I",
    "user": "alice",
    "time": "2020-06-12T00:43:26Z",
    // ...
}
```

ollyg commented:
just to note that I think the extension is permit-X11-forwarding (uppercase X) and this caught me out! thank you so much for working on this 😄

fspmarshall committed:
proxy X11 forwarding support

- Role options now include a `permit_x11_forwarding` bool
which is set to `false` by default.

- Recording proxies now forward X11 requests and channels
when when permitted by RBAC.

- User certs will now include the `permit-X11-forwarding`
extension when permitted by RBAC.

- If X11 forwarding is requested for a session a new `x11`
audit event is emitted by recording proxies.

awly commented on lib/srv/forward/sshserver.go:
```diff
@@ -861,6 +864,182 @@ func (s *Server) handleAgentForward(ch ssh.Channel, req *ssh.Request, ctx *srv.S
+		select {
+		case sreq := <-sin:
+			if sreq == nil {
+				return nil
```
Do we need to return right away on channel closure?
If not, how about "disabling" the select case for a closed channel (setting `sin` to `nil`) in this case?
This will make it so that the function only returns after both `io.Copy` goroutines have finished, which in turn makes those `sync.Once`s unnecessary.

awly commented on lib/srv/forward/sshserver.go:
```diff
@@ -861,6 +864,182 @@ func (s *Server) handleAgentForward(ch ssh.Channel, req *ssh.Request, ctx *srv.S
+		}
+	}()
+
+	// Emit a x11 forwarding audit event.
+	s.EmitAuditEvent(events.X11Forward, events.EventFields{
+		events.EventLogin: s.identityContext.Login,
+		events.EventUser:  s.identityContext.TeleportUser,
+		events.LocalAddr:  s.sconn.LocalAddr().String(),
+		events.RemoteAddr: s.sconn.RemoteAddr().String(),
+	})
```
Also emit an event on any failure (permission check or `forwardRequest` can fail)

awly commented on lib/srv/forward/sshserver.go:
```diff
@@ -861,6 +864,182 @@ func (s *Server) handleAgentForward(ch ssh.Channel, req *ssh.Request, ctx *srv.S
+func (s *Server) serveX11Channels(ctx context.Context) error {
+	ncs := s.remoteClient.HandleChannelOpen(sshutils.X11ChannelRequest)
+	if ncs == nil {
+		return trace.AlreadyExists("x11 forwarding already active")
```
nit: `s/x11/X11` to be consistent

fspmarshall replied:
I'm deliberately using `x11` instead of `X11` in places where proper nouns are typically lowercase (e.g. error messages, config fields).  This is consistent with how openssh handles it, and with how teleport generally treats terms like `http`/`HTTP`.

awly reviewed (commented):
Please add unit and/or integration tests

russjones commented on lib/events/api.go:
```diff
@@ -206,6 +206,9 @@ const (
 	SubsystemError = "exitError"
 
+	// X11 forwarding event
+	X11ForwardEvent = "x11"
```
Let's give this a more descriptive name, maybe `x11-forwarding`?

russjones commented on lib/srv/forward/sshserver.go:
```diff
@@ -861,6 +864,182 @@ func (s *Server) handleAgentForward(ch ssh.Channel, req *ssh.Request, ctx *srv.S
+		cch.Close()
+	}()
+
+	go func() {
+		// forward data from server to client
+		io.Copy(cch, sch)
+		// inform client that no more data is coming
+		ccw.Do(func() {
+			cch.CloseWrite()
+		})
+		// if we are the second io forwarder to complete,
+		// then signal shutdown to main goroutine.
+		if atomic.SwapUint32(&ioflag, 1) > 0 {
+			cancel()
+		}
+	}()
```
Can we simplify the logic here? We have a lot of these `io.Copy` loops in our code base, ideally we'd follow the same pattern everywhere.

fspmarshall commented on lib/services/types.proto:
```diff
@@ -685,6 +685,13 @@ message RoleOptions {
     repeated string BPF = 7 [ (gogoproto.jsontag) = "enhanced_recording,omitempty" ];
+
+    // PermitX11Forwarding authorizes use of X11 forwarding.
+    bool PermitX11Forwarding = 8 [
+        (gogoproto.nullable) = true,
```
Pretty sure there isn't an `optional` in `proto3`.

awly commented on lib/srv/forward/sshserver.go:
```diff
@@ -861,6 +863,184 @@ func (s *Server) handleAgentForward(ch ssh.Channel, req *ssh.Request, ctx *srv.S
+	// setup wait group for io forwarding goroutines
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		// forward data from server to client
+		io.Copy(cch, sch)
+		// inform client that no more data is coming
+		cch.CloseWrite()
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		// forward data from client to server
+		io.Copy(sch, cch)
+		// inform server that no more data is coming
+		sch.CloseWrite()
+	}()
```
Nothing calls `cch.Close()` and `sch.Close()`, are these `CloseWrite` calls sufficient to clean up the connection?

fspmarshall replied:
Good catch, `CloseWrite` is not sufficient.  Accidentally dropped the `Close` call when refactoring.

awly commented on lib/srv/regular/sshserver_test.go:
```diff
@@ -1148,6 +1149,228 @@ func (s *SrvSuite) TestGlobalRequestRecordingProxy(c *C) {
+		return nil, nil, nil, trace.Wrap(err)
+	}
+	srvConn, chs, reqs, err := ssh.NewServerConn(netConn, &r.cfg)
+	if err != nil {
+		return nil, nil, nil, trace.Wrap(err)
+	}
```
nit: `netConn.Close()` in case of error

awly commented on lib/srv/regular/sshserver_test.go:
```diff
@@ -1148,6 +1149,228 @@ func (s *SrvSuite) TestGlobalRequestRecordingProxy(c *C) {
+// to it.  Used to verify the behavior of X11 forwarding in recording proxies.
+func (s *SrvSuite) startX11EchoServer(ctx context.Context, c *C) *rawNode {
+	node := s.newRawNode(c)
+	ctx, cancel := context.WithCancel(ctx)
```
nit: move this inside the goroutine

awly commented on lib/srv/regular/sshserver_test.go:
```diff
@@ -1148,6 +1149,228 @@ func (s *SrvSuite) TestGlobalRequestRecordingProxy(c *C) {
+				case <-ctx.Done():
+					return
+				}
+				c.Assert(nch.ChannelType(), Equals, teleport.ChanSession)
```
I believe `c.Assert` is not thread-safe.
Probably easier to keep as is, but the race detector might complain.

awly commented on lib/srv/regular/sshserver_test.go:
```diff
@@ -1148,6 +1149,228 @@ func (s *SrvSuite) TestGlobalRequestRecordingProxy(c *C) {
+
+				// expect client to open a session channel
+				var nch ssh.NewChannel
+				select {
+				case nch = <-chs:
+				case <-ctx.Done():
+					return
+				}
```
for this and next selects, add a timeout.
it'll let the test fail faster if client is mis-behaving.

awly commented on lib/srv/regular/sshserver_test.go:
```diff
@@ -1148,6 +1149,228 @@ func (s *SrvSuite) TestGlobalRequestRecordingProxy(c *C) {
+				}
+				c.Assert(nch.ChannelType(), Equals, teleport.ChanSession)
+
+				_, creqs, err := nch.Accept()
```
nit: don't forget to close the `ssh.Channel` here

awly commented on lib/srv/regular/sshserver_test.go:
```diff
@@ -1148,6 +1149,228 @@ func (s *SrvSuite) TestGlobalRequestRecordingProxy(c *C) {
+				defer xch.Close()
+				// echo all bytes back across the X11 channel
+				if _, err := io.Copy(xch, xch); err != nil {
+					if err == io.EOF {
```
`io.Copy` should never return `EOF`

awly commented on lib/srv/regular/sshserver_test.go:
```diff
@@ -1148,6 +1149,228 @@ func (s *SrvSuite) TestGlobalRequestRecordingProxy(c *C) {
+	return node
+}
+
+func writeAll(dst io.Writer, buf []byte) error {
+	w := 0
+	for w < len(buf) {
+		n, err := dst.Write(buf[w:])
+		if err != nil {
+			return err
+		}
+		w += n
+	}
+	return nil
+}
```
`io.Writer` should always return a non-nil error on a short write.
this helper func likely behaves just like `dst.Write(buf)`

fspmarshall replied:
Woah, I always assumed it returned after the first successful write... cool!

awly commented on lib/srv/regular/sshserver_test.go:
```diff
@@ -1148,6 +1149,228 @@ func (s *SrvSuite) TestGlobalRequestRecordingProxy(c *C) {
+	node := s.startX11EchoServer(ctx, c)
+
+	// Create a direct TCP/IP connection from proxy to our X11 test server.
+	netConn, err := s.clt.Dial("tcp", node.addr)
```
don't forget to close this conn

sjackman commented:
```console
$ tsh version  
Teleport v7.3.3 git: go1.17.2
$ tsh ssh -X remote-host
ERROR: unknown short flag '-X'
```
How do I use X11 forwarding with `tsh ssh`?

ollyg commented:
Hi @sjackman you can only use -X with the openssh client (so, using `ssh` instead of `tsh ssh`). There are other restrictions such as not using recording mode on the proxy, best to get with Gravitational support for the details.
